continuationsPlugin/src/main/scala/continuations/ContinuationsPlugin.scala

```diff
@@ -74,3 +96,141 @@ class ContinuationsPhase extends PluginPhase:
     tree match
       case Apply(_, _) => returnsContextFunctionWithSuspendType(tree)
       case _ => false
+
+  def hasOnlySuspendParam(tree: DefDef)(using Context): Boolean =
+    tree.termParamss match
+      case List(Nil, param :: Nil) => isSuspendType(param.tpe)
```

Comment:
This will work on *our* lone example. But it would fail if there are more using params, and it would accidentally trigger the rewrite in something like this:

```scala
def myNonSupsendFunction(s: Suspend): = ???
```

To check if tree is a `DefDef`, and if it has a `using Suspend` parameter, we need to check `tree.isDef` is `true` cast it as a `tpd.DefDef`, then find if a parameter in its `paramss` exists that is `Flags.Given` and that the parameter's type is `continuations.Suspend`.

Here's a working example from jackcviers/TBD:30m3gh3-non-suspending-0-arity local work progress with tests and results.

```scala
package continuations

import dotty.tools.dotc.ast.Trees
import dotty.tools.dotc.ast.tpd
import dotty.tools.dotc.core.Contexts.Context
import dotty.tools.dotc.core.Flags
import dotty.tools.dotc.report

object HasSuspendParameter {

  def unapply(tree: tpd.Tree)(using c: Context): Option[tpd.Tree] = Option(tree).filter{t =>
    t.isDef && t.asInstanceOf[tpd.DefDef].paramss.exists{ pc =>
      pc.exists{ v => v.denot.symbol.is(Flags.Given) && v.tpe.classSymbol.showFullName == "continuations.Suspend" }
    }
  }
}
```
Tests:
```scala
package continuations

import dotty.tools.dotc.core.Contexts.Context
import dotty.tools.dotc.core.Symbols
import dotty.tools.dotc.core.Types.Type
import dotty.tools.dotc.ast.tpd
import munit.FunSuite
import dotty.tools.dotc.core.Names
import dotty.tools.dotc.core.Flags
import dotty.tools.dotc.core.Constants.Constant.apply
import dotty.tools.dotc.core.Constants
import dotty.tools.dotc.ast.Trees

class HasSuspendParmeterSuite extends FunSuite, CompilerFixtures {

  compilerContextWithContinuationsPlugin.test(
    "#unapply(defDefTree): def mySuspend()(using Suspend): Int = 1 should be Some(tree)") {
    case given Context =>
      val c = summon[Context]
      val suspend = Symbols.requiredClass("continuations.Suspend")
      val d = tpd.DefDef(
        Symbols
          .newSymbol(
            c.owner,
            Names.termName("mySuspend"),
            Flags.EmptyFlags,
            c.definitions.IntType)
          .asTerm,
        List(
          List(),
          List(
            Symbols.newSymbol(
              c.owner,
              Names.termName("x$1"),
              Flags.union(Flags.GivenOrImplicit, Flags.Param),
              suspend.info))),
        c.definitions.IntType,
        tpd.Literal(Constants.Constant(1))
      )
      assertEquals(HasSuspendParameter.unapply(d), Some(d))
  }

  compilerContextWithContinuationsPlugin.test(
    "#unapply(defDefTree): def mySuspend(s: Suspend): Int = 1 should be None") {
    case given Context =>
      val c = summon[Context]
      val suspend = Symbols.requiredClass("continuations.Suspend")
      val d = tpd.DefDef(
        Symbols
          .newSymbol(
            c.owner,
            Names.termName("mySuspend"),
            Flags.EmptyFlags,
            c.definitions.IntType)
          .asTerm,
        List(
          List(
            Symbols.newSymbol(
              c.owner,
              Names.termName("s"),
              Flags.union(Flags.Param),
              suspend.info))),
        c.definitions.IntType,
        tpd.Literal(Constants.Constant(1))
      )
      assertEquals(HasSuspendParameter.unapply(d), None)
  }
}
```
Results:
```scala
continuations.HasSuspendParmeterSuite:
  + #unapply(defDefTree): def mySuspend()(using Suspend): Int = 1 should be Some(tree) 0.458s
  + #unapply(defDefTree): def mySuspend(s: Suspend): Int = 1 should be None 0.053s
[info] Passed: Total 2, Failed 0, Errors 0, Passed 2
[success] Total time: 2 s, completed Nov 4, 2022, 12:19:15 PM
```
.

Reply:
💡 If we create a class like suggested above, we can move this method there (and avoid possible merge conflicts).

Reply:
added the more general case https://github.com/47deg/TBD/pull/52/commits/1a5f9ef8011ed082a480c0da6afa4bae1cc71944#diff-fade65ebf959155389570d62d274801c2f7607f965a273661e5f776636d6393fR131

Reply:
Yes. Looks good.